pkgtools/import-package/files/import-package.1
.\"	$NetBSD: import-package.1,v 1.2 2020/08/16 20:57:57 wiz Exp $
.\"
.Dd August 16, 2020
.Dt IMPORT_PACKAGE 1
.Sh NAME
.Nm import-package
.Nd import package to pkgsrc
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
.Nm
is a script designed to make adding packages into pkgsrc or pkgsrc-wip
easier.
To use it, just
.Ic cd
to the package directory and run
.Nm .
.Pp
.Nm
will automatically create a nice import message based on
.Pa DESCR
and
.Va PKGNAME ,
set up the CVS tags correctly and autodetect what CVSROOT (or git)
to use.
.Nm
also shows you what files will be imported, reminds you
to run
.Xr pkglint 1
and asks for confirmation before doing anything.
.Sh AUTHORS
.An Tobias Nygren
.An Filip Hajny
.An Thomas Klausner
.An dieter roelants
.An Kamil Rytarowski
.An Ryo ONODERA
.Sh CAVEATS
.Nm
does not create the
.Pa CHANGES-*
entry, so please run
.Dl "PKG_DEVELOPER=yes make cce CTYPE=Added"
after the import, if importing into pkgsrc.
Nor does
.Nm
add the package directory to the category's
.Pa Makefile ,
so please don't forget to do that either.


doc/NEWS
# $NetBSD: NEWS,v 1.13 2020/03/30 10:04:27 nia Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"
	to invoke this functionality. Set PKG_DESTDIR_SUPPORT to "none"
	to revert to direct installation.

buildlink3.mk layout changed [joerg 2009-03-20]
	The buildlink3.mk files have been reorganized to reduce
	redundant parsing for deeply nested buildlink3 trees.
	The createbuildlink package has been updated to create
	the new format.

PLIST_VARS introduced [jlam 20080413]
	PLIST_VARS is a simpler way to do conditional PLIST entries.
	The old method of doing e.g. PLIST_SUBST+= FOO="@comment " can
	now be done by unconditionally setting PLIST_VARS+=foo and
	conditionally setting PLIST.foo=yes. This will result in substitution
	of ${PLIST.foo} in the PLIST with either "" or "@comment ".

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


pkgtools/import-package/files/import-package.sh
#! /bin/sh
#
# $NetBSD: import-package.sh,v 1.1 2020/08/16 20:57:09 wiz Exp $
#
# Script designed to make add packages into wip and main pkgsrc easier.
#
# Just cd to the package directory and run import-package
#
# It will automatically create a nice import message based on DESCR
# and PKGNAME, set up the CVS tags correctly and autodetect what CVSROOT
# to use. It also shows you what files will be imported, reminds you
# to run pkglint(1) and asks for confirmation before doing anything.

set -e

[ -n "${MKTEMP}" ] || MKTEMP=mktemp
[ -n "${EDITOR}" ] || EDITOR=vi
CLEANUP=""
DRYRUN="" # "echo dry-run:"

cleanup() {
	if [ -n "${CLEANUP}" ]; then
		rm -f ${CLEANUP}
	fi
}
trap cleanup 0

if [ -z "${MAKE}" ]; then
	if type bmake >/dev/null 2>&1; then
		MAKE=bmake
	else
		MAKE=make
	fi
fi

if [ ! -f "../../mk/bsd.pkg.mk" ]; then
	echo "$0: ../../bsd.pkg.mk not found" 1>&2
	exit 1
fi

stale=no
for cvsdir in $(find "$(pwd)" -type d -name CVS -print); do
	echo "$0: stale CVS state directory found: ${cvsdir}" 1>&2
	stale=yes
done
for gitdir in $(find "$(pwd)" -type d -name .git -print); do
	echo "$0: stale git state directory found: ${gitdir}" 1>&2
	stale=yes
done
for wrkdir in $(find "$(pwd)" -type d -name "work*" -print); do
	echo "$0: stale work directory found: ${wrkdir}" 1>&2
	stale=yes
done
if [ "${stale}" = "yes" ]; then
	exit 1
fi

PACKAGE="$(basename $(pwd))"
CATEGORY="$(basename $(dirname $(pwd)))"
PKGPATH="${CATEGORY}/${PACKAGE}"

if [ "${CATEGORY}" = "wip" ]; then
	SCM=GIT
	if [ ! -d "../.git" ]; then
		echo "$0: git state directory not found in ../.git" 1>&2
		exit 1
	fi
	GIT_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
	if [ "${GIT_BRANCH}" != "master" ]; then
		echo "$0: ${GIT_BRANCH} is not the git master branch" 1>&2
		exit 1
	fi
	if ! git config --get user.name > /dev/null; then
		echo "$0: Please run 'git config --local user.name \"Your name here\"'" 1>&2
		exit 1
	fi
	if ! git config --get user.email > /dev/null; then
		echo "$0: Please run 'git config --local user.email \"you@example.com\"'" 1>&2
		exit 1
	fi
	if git status | grep "Changes to be committed" > /dev/null; then
		echo "$0: You have uncommitted changes staged!" 1>&2
		echo "$0: Check output of 'git status'" 1>&2
		exit 1
	fi
	if ! git fetch; then
	    echo "$0: git fetch failed"
	fi
	if ! git status -u no | egrep "Your branch is (up-to-date|up to date)" > /dev/null; then
		echo "$0: git repository does not appear to be up to date." 1>&2
		echo "$0: You should probably run 'git pull --rebase'" 1>&2
		echo "$0: and 'git push' (if you have unpushed work)" 1>&2
		exit 1
	fi
else
	SCM=CVS
	CVSROOT="$(cat ../CVS/Root | tr A-Z a-z | sed -e 's/.*@//')"
	if [ "${CVSROOT}" != "cvs.netbsd.org:/cvsroot" ]; then
		echo "$0: wrong CVS root" 1>&2
		exit 1
	fi
fi

MSG="$(${MKTEMP} -t import-package.XXXXXXXX)"
CLEANUP="${MSG}"
echo "Please wait while determining PKGNAME and DESCR_SRC."
PKGNAME="$(${MAKE} show-var VARNAME=PKGNAME)"
DESCR_SRC="$(${MAKE} show-var VARNAME=DESCR_SRC) /dev/null"

DASH70=----------------------------------------------------------------------

echo "${CATEGORY}/${PACKAGE}: import ${PKGNAME}" > ${MSG}
echo "" >> ${MSG}
cat ${DESCR_SRC} >> ${MSG}
echo "${SCM}: ${DASH70}" >> ${MSG}
echo "${SCM}: Please edit the above message to give a brief description" >> ${MSG}
echo "${SCM}: of the package for those who read the *-changes@ list." >> ${MSG}
echo "${SCM}: Did you remember to run pkglint(1) before importing?" >> ${MSG}
echo "${SCM}:" >> ${MSG}
echo "${SCM}: Lines starting with ${SCM}: will be automatically removed." >> ${MSG}
echo "${SCM}:" >> ${MSG}

ADDLIST="$(${MKTEMP} -t import-package-files.XXXXXXXX)"
CLEANUP="${CLEANUP} ${ADDLIST}"
(
	cd ..
	if [ "${SCM}" = "GIT" ]; then
		find ${PACKAGE} -type f -print
	else
		find ${PACKAGE} \( -type d \) -print | sed 's,$,/,'
		find ${PACKAGE} \( -type f \) -print
	fi
) | sed -e '/^\./d' -e '/\/CVS/d' -e '/[^+-_,./a-zA-Z0-9]/d' | sort > ${ADDLIST}
sed "s|^|${SCM}: will add: ${CATEGORY}/|" ${ADDLIST} >> ${MSG}

${EDITOR} ${MSG}

echo "Edited message follows:"
echo ${DASH70}
grep -v "^${SCM}:" < ${MSG}
echo ${DASH70}
echo 	"PKGPATH:	${PKGPATH}"
if [ "${SCM}" = "GIT" ]; then
	echo 	"GIT_BRANCH:	${GIT_BRANCH}"
else
	echo 	"CVSROOT:	${CVSROOT}"
fi

echo ""
printf "y + enter to import, any other text + enter to abort> "
read ANS

if [ "${ANS}" = "y" ]; then
	if [ "${SCM}" = "GIT" ]; then
		(
			cd ..
			cat ${ADDLIST} | xargs -L 100 ${DRYRUN} git add
			${DRYRUN} git commit -m "$(grep -v "^${SCM}:" ${MSG})" ${PACKAGE}
			${DRYRUN} git push
		)
	else
		(
			cd ..
			export CVS_RSH=ssh
			[ -e "${PACKAGE}/CVS" ] || ${DRYRUN} cvs add ${PACKAGE} || exit 1
			grep '/$' ${ADDLIST} | fgrep -vx ${PACKAGE}/ |
			    xargs -L 100 ${DRYRUN} cvs add
			grep -v '/$' ${ADDLIST} | xargs -L 100 ${DRYRUN} cvs add
			${DRYRUN} cvs commit -m "$(grep -v "^${SCM}:" ${MSG})" ${PACKAGE}
		)
	fi
	echo ${DASH70}
	echo "Don't forget to add the package to ${CATEGORY}/Makefile."
	echo "When imported to pkgsrc itself, please update the CHANGES-*"
	echo "file and possibly remove the package from the TODO list."
	echo ""
fi
